private/bufpkg/bufrpc/interceptors.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufrpc
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/netrc"
+	"github.com/bufbuild/connect-go"
+)
+
+const (
+	// AuthenticationHeader is the standard OAuth header used for authenticating
+	// a user. Ignore the misnomer.
+	AuthenticationHeader = "Authorization"
+	// AuthenticationTokenPrefix is the standard OAuth token prefix.
+	// We use it for familiarity.
+	AuthenticationTokenPrefix = "Bearer "
+)
+
+const (
+	tokenEnvKey = "BUF_TOKEN"
+)
+
+// NewWithTokenReaderInterceptor returns a new Connect Interceptor that looks up an auth token on every request and when
+// found, sets it into the request header if not already set.
+func NewWithTokenReaderInterceptor(container appflag.Container, address string) connect.UnaryInterceptorFunc {
+	interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
+		return connect.UnaryFunc(func(
+			ctx context.Context,
+			req connect.AnyRequest,
+		) (connect.AnyResponse, error) {
+			token := container.Env(tokenEnvKey)
+			if token == "" {
+				machine, err := netrc.GetMachineForName(container, address)
```

Comment:
I'm not sure what we were doing before but we definitely don't want to do this for every RPC, this is a file-based operation on a per-RPC basis

Reply:
This is per-RPC call yeah.  We could move this to the client level so that the token is looked up when the client is created.

Reply:
Yea I think we were doing this at CLI instantiation effectively

Reply:
Just to clarify too -- this is how it's behaving currently also.  The context modifier gets run on every RPC invocation.

Reply:
So it is - just ran through it.